src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs

```diff
@@ -197,7 +198,11 @@ protected static JsonConverter<TKey> GetKeyConverter(Type keyType, JsonSerialize
                             }
                         }
 
-                        state.Current.DictionaryKey = ReadDictionaryKey(ref reader, ref state);
+                        key = ReadDictionaryKey(ref reader, ref state);
+                    }
+                    else
+                    {
+                        key = (TKey)state.Current.DictionaryKey!;
```

Comment:
Why can we do this? Also, what's guaranteeing that `DictionaryKey` isn't null here?

Reply:
> This is great. How did you detect it?
> 
> > In the best-case scenario when an entire dictionary fits in a buffer - it should completely remove unnecessary heap allocations for keys.
> 
> Can we run some benchmarks to get a sense of how much heap allocations we save? It might help modifying and running this benchmark on a dictionary input that you expect to get performance wins on (with value type keys). Please share the benchmarkdotnet results when you have them:
> https://github.com/dotnet/performance/blob/0b094042e7dc114c8cd038b1ae90dad0e5696fa2/src/benchmarks/micro/libraries/System.Text.Json/Serializer/ReadJson.cs#L57-L64

I am currently working on the high-performance dynamodb library which involves some low-level optimized JSON deserialization. I was able to detect several inefficiencies during some internal benchmark comparisons.

Here is the benchmark result for `Dictionary<int, string>` with `100` items. I think the most important column for this case is `Allocated`, where we can see that the `2KB` difference disappears after the PR. It also becomes slightly faster.

``` ini

BenchmarkDotNet=v0.12.1.1466-nightly, OS=Windows 10.0.19041.572 (2004/May2020Update/20H1)
Intel Core i7-7820X CPU 3.60GHz (Kaby Lake), 1 CPU, 6 logical and 3 physical cores
.NET SDK=6.0.100-alpha.1.20630.6
  [Host]     : .NET 6.0.0 (6.0.20.62903), X64 RyuJIT
  Job-NDNXAQ : .NET 6.0.0 (42.42.42.42424), X64 RyuJIT
  Job-EVXCXL : .NET 6.0.0 (42.42.42.42424), X64 RyuJIT

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:DebugType=portable  IterationTime=250.0000 ms  
MaxIterationCount=20  MinIterationCount=15  WarmupCount=1  

```
|                   Method |        Job |                                                                                                                Toolchain |     Mean |    Error |   StdDev |   Median |      Min |      Max | Ratio | RatioSD |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|------------------------- |----------- |------------------------------------------------------------------------------------------------------------------------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|--------:|-------:|-------:|------:|----------:|
|    DeserializeFromString | Job-NDNXAQ | master | 19.96 μs | 0.156 μs | 0.138 μs | 19.93 μs | 19.77 μs | 20.20 μs |  1.00 |    0.00 | 3.2861 | 0.2347 |     - |     19 KB |
|    DeserializeFromString | Job-EVXCXL |     PR | 20.05 μs | 0.304 μs | 0.312 μs | 19.94 μs | 19.77 μs | 20.84 μs |  1.01 |    0.02 | 3.2727 | 0.1596 |     - |     19 KB |
|                          |            |                                                                                                                          |          |          |          |          |          |          |       |         |        |        |       |           |
| DeserializeFromUtf8Bytes | Job-NDNXAQ | master | 19.56 μs | 0.232 μs | 0.194 μs | 19.54 μs | 19.30 μs | 19.97 μs |  1.00 |    0.00 | 3.2396 | 0.1580 |     - |     19 KB |
| DeserializeFromUtf8Bytes | Job-EVXCXL |     PR | 19.52 μs | 0.138 μs | 0.122 μs | 19.52 μs | 19.34 μs | 19.78 μs |  1.00 |    0.01 | 3.2233 | 0.1572 |     - |     19 KB |
|                          |            |                                                                                                                          |          |          |          |          |          |          |       |         |        |        |       |           |
|    DeserializeFromStream | Job-NDNXAQ | master | 21.66 μs | 0.187 μs | 0.175 μs | 21.58 μs | 21.49 μs | 22.17 μs |  1.00 |    0.00 | 3.6509 | 0.2608 |     - |     21 KB |
|    DeserializeFromStream | Job-EVXCXL |     PR | 21.25 μs | 0.201 μs | 0.178 μs | 21.27 μs | 20.90 μs | 21.57 μs |  0.98 |    0.01 | 3.2669 | 0.1675 |     - |     19 KB |

It is also worth noting another potential optimization that I am considering to create a PR for. On the benchmark, you can see that deserialization from `Stream` is visibly slower than from bytes array even when entire JSON fits in a single buffer. I suspect it happens because we don't use `UseFastPath` for a single buffer case. But it is unrelated and probably not worth discussing in this thread.

Reply:
In a similar way to other converters like `ObjectDefaultConverter`, `else` statement is needed for the "resume" scenario when after reading the key buffer ends (we return `false`). Later on, when the next buffer is read, we rebuild the call stack, reach the current method, and this time `state.Current.PropertyState >= StackFramePropertyState.Name`. 

>  can you infer what's guaranteeing that DictionaryKey isn't null here?
`else` statement should be unreachable in case when key is null, because we only access it when `state.Current.PropertyState >= StackFramePropertyState.Name` (meaning we access it only when it is initialized). It is the reason why we assign `state.Current.DictionaryKey = key` for all `return false` cases.

Reply:
> `else` statement should be unreachable in case when key is null, because we only access it when `state.Current.PropertyState >= StackFramePropertyState.Name` (meaning we access it only when it is initialized). It is the reason why we assign `state.Current.DictionaryKey = key` before all `return false` cases.

Thanks. This is useful context that would be helpful as a comment on this line for future code readers. Consider adding it.